File: go800mon/a800mon/cpustate.go
package a800mon

import "context"

type CpuStateViewer struct {
	BaseWindowComponent
	lastSnapshot string
}

func NewCpuStateViewer(window *Window) *CpuStateViewer {
	v := &CpuStateViewer{BaseWindowComponent: NewBaseWindowComponent(window)}
	return v
}

func (v *CpuStateViewer) Update(_ctx context.Context) (bool, error) {
	st := State()
	snap := formatCPU(st.CPU) + "|" + st.CPUDisasm
	if v.lastSnapshot == snap {
		return false, nil
	}
	v.lastSnapshot = snap
	return true, nil
}

func (v *CpuStateViewer) Render(_force bool) {
	st := State()
	line := formatCPU(st.CPU)
	if st.CPUDisasm != "" {
		line += "  " + st.CPUDisasm
	}
	w := v.Window()
	w.Cursor(0, 0)
	w.Print(line, ColorText.Attr(), false)
	w.ClearToEOL(false)
}

File: py800mon/monitor/cpustate.py
from ..app import Component, VisualComponent
from .appstate import state


class CpuStateViewer(VisualComponent):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._last_snapshot = None

    async def update(self):
        snapshot = (state.cpu, state.cpu_disasm)
        if self._last_snapshot == snapshot:
            return False
        self._last_snapshot = snapshot
        return True

    def render(self, force_redraw=False):
        line = repr(state.cpu)
        if state.cpu_disasm:
            line += f"  {state.cpu_disasm}"
        self.window.print(line)
        self.window.clear_to_eol()
